Make sure exit code != 0 if incompatible archive version

CasperWA opened the pull request:
Fixes #2633.

For `verdi import`:
Change the warning to a critical if exception "IncompatibleArchiveVersionError" is raised.

Added test `test_import_old_versions`.
It checks a non-zero value is returned, when an archives with old versions are imported.
Also, added asserts of exit code for other `verdi import` tests.

sphuber commented on aiida/cmdline/commands/cmd_import.py:
```diff
@@ -127,7 +127,7 @@ def cmd_import(archives, webpages, group, extras_mode_existing, extras_mode_new,
         except exceptions.IncompatibleArchiveVersionError as exception:
-            echo.echo_warning('{} cannot be imported: {}'.format(archive, exception))
+            echo.echo_critical('{} cannot be imported: {}'.format(archive, exception))
             continue
```
Since the `echo_critical` call will call `sys.exit` you can remove the `continue`

CasperWA replied:
Done.

CasperWA commented:
I noticed I missed this correction for an archive URL, this was only done for a local archive.
I have corrected this initial oversight and removed `continue` in both places.
If you have a valid URL to an archive, please paste it and I will add another test for this if needed.

ltalirz commented:
@CasperWA well, I have lots of valid urls for archives (on the materials cloud archive) but those are all archives created by aiida-core 0.12.
I think it's important that we add a flag to `verdi import` that does the migration automatically (but not in this PR).

CasperWA commented:
After discussing it over with @ltalirz, I will import the same archives, but from GitHub, referencing a specific commit that contains the files.
There is still the problem of running the tests if one has no internet connection, however, specifically for the "make sure it fails"-test, this shouldn't be a problem, other than the fact we're not testing _exactly_ what we want to test...

CasperWA commented:
`verdi import` is broken with respect to URLs.

I will rewrite it in this PR, since it is essentially fixing the code to make the tests work and get the result that is sought after in issue #2633.

CasperWA commented:
> I will rewrite it in this PR, ...

Done :)

ltalirz commented on aiida/backends/tests/cmdline/commands/test_import.py:
```diff
@@ -85,8 +88,67 @@ def test_comment_mode(self):
+        """
+        url = "https://raw.githubusercontent.com/aiidateam/aiida_core/f685844b290ed23df254873df62da7162bde0f1b/"
+        archive_path = "aiida/backends/tests/fixtures/export/migrate/"
+        archives = [('export_v0.1.aiida', '0.1'), ('export_v0.2.aiida', '0.2'), ('export_v0.3.aiida', '0.3')]
```
Fine to test the import via URL once; no need to do requests for all versions (it's already tested above)

CasperWA replied:
Sure, will reduce.

ltalirz commented on aiida/backends/tests/cmdline/commands/test_import.py:
```diff
@@ -85,8 +88,67 @@ def test_comment_mode(self):
+        self.assertEqual(result.exit_code, 0, result.output)
+
+    def test_import_url_timeout(self):
+        """Test a timeout to valid URL is correctly errored"""
```
hm... this I find a but problematic.
the rule of thumb for unit tests is that they should run "in the blink of an eye".
are you waiting for a timeout here? is there a way to reduce this test to ~100ms?

CasperWA replied:
Yes, I agree about this. But I want/need to test that we can get to this point in the code. If you have a better suggestion I'm happy to hear it. The timeout is otherwise "only" 10 seconds.

CasperWA replied:
I've thought of something. I'll make `timeout` a property exactly for this purpose. Then we can test it with `timeout = 0`.

ltalirz commented on aiida/cmdline/commands/cmd_import.py:
```diff
@@ -25,6 +26,52 @@
+    """AiiDA Verdi implementation of Click's Path-type.
+    Introduce tests to distinguish whether the path is to a local file
+    or a URL string to a file that needs to be downloaded.
+    """
```
Suggestion:

```
"""AiiDA extension of Click's Path-type to include URLs.

An ImportPath can either be a Click Path-type or a URL.
"""
```

CasperWA replied:
Done.

ltalirz commented on aiida/cmdline/commands/cmd_import.py:
```diff
@@ -25,6 +26,52 @@
+        super(ImportPath, self).__init__(**kwargs)
+
+    def __call__(self, value, param=None, ctx=None):
+        if value is not None:
```
is that check necessary here?
shouldn't we let covert handle that?

CasperWA replied:
I'm simply reimplementing the `__call__` function as it is for `click.Path`'s parent class `ParamType`. I think it's fine to handle it here and let `convert` handle the more detailed checks.

ltalirz commented on aiida/cmdline/commands/cmd_import.py:
```diff
@@ -25,6 +26,52 @@
+    def __call__(self, value, param=None, ctx=None):
+        if value is not None:
+            try:
+                res = super(ImportPath, self).convert(value, param, ctx)
```
since you are reusing the `convert` - perhaps it would be cleaner to simply overwrite `convert` directly?

i.e.
```
def convert(..):
   try:
       return super(ImportPath, self).convert(value, param, ctx)
    except click.exceptions.BadParameter:
       new code...
```

the way you do it currently, if somebody were to use your new `convert` method, the result would not be the expected one (since it only handles the URL case)

CasperWA replied:
True, This way I don't have to overwrite the `__call__` function as well.

CasperWA replied:
Done

ltalirz commented on aiida/cmdline/commands/cmd_import.py:
```diff
@@ -25,6 +26,52 @@
+        from socket import timeout
+
+        url = value
+        timeout_seconds = 10
```
this should probably go as a constant in some configuration module

CasperWA replied:
Probably should. And I think the whole class should go somewhere else as well. I was kind of hoping @sphuber could help me out on this :)

CasperWA replied:
I have moved `ImportPath` to `aiida.cmdline.params.path`.
The timeout seconds are moved into this file as well.

ltalirz commented on aiida/cmdline/commands/cmd_import.py:
```diff
@@ -25,6 +26,52 @@
+        try:
+            URLValidator()(url)
+        except URLValidationError:
+            self.fail('%s "%s" is not a valid URL or local %s.' % (self.path_type, filename_to_ui(url), self.name),
```
why "or local"?

CasperWA replied:
Because if it gets to this point, we simply assume that the input is a URL, but it may just be a mistyped local path. So to not completely confuse the user, i put in "or local".

ltalirz replied:
You're right. In this case, what about the following error message:
```
'"%s" is neither a valid filepath nor a valid URL.' % (filename_to_ui(url))
```

ltalirz replied:
if you want to make things more "analogous" between checking whether it is a path or a URL, you can do a 
```
try:
   ...
except: 
    pass
```
for both and at the very end of the function fail with this error message.

CasperWA replied:
That's a nice idea.

ltalirz commented on aiida/cmdline/commands/cmd_import.py:
```diff
@@ -159,8 +205,7 @@ def cmd_import(archives, webpages, group, extras_mode_existing, extras_mode_new,
                 echo.echo_warning('{} cannot be imported: {}'.format(archive, exception))
-                echo.echo_warning('download the archive file and run `verdi export migrate` to update it')
-                continue
+                echo.echo_critical('download the archive file and run `verdi export migrate` to update it')
```
since these are not two independent messages, probably  better to construct the string first and then pass it to `echo_critical` directly

CasperWA replied:
I agree, I thought to do this as well, but forgot it again as quickly, unfortunately. Good catch.

ltalirz commented:
Thanks for looking into this!
Through with my review

sphuber commented on aiida/cmdline/params/types/path.py:
```diff
@@ -31,3 +33,76 @@ def convert(self, value, param, ctx):
+    # pylint: disable=attribute-defined-outside-init
+    @timeout_seconds.setter
+    def timeout_seconds(self, value):
+        if isinstance(value, str):
```
This whole block can be cleaned up. First off, you can get either a `ValueError` or `TypeError` if the type of the timeout is incorrect. Furthermore, the combination of `if` conditional and then try-except is superfluous and not very pythonic. I would change this whole block simply in:
```
try:
     self._timeout_seconds = int(value)
except ValueError:
    raise TypeError('timeout should be an integer but got: {}'.format(type(value)))
```
The default should simply be set in the constructor of this object.

CasperWA replied:
Done. I have reduced it to your more simple try-except block, keeping the following limit checks.
The default value is set in the constructor with a check `if timeout_seconds is None` set it to the default.
I think it is maybe better to keep a None value in the `__init__` parameters, since if the default is set already here and someone should pass a `None` value the default value will still be used. Unless we specifically want it to raise if a None value is passed?

sphuber commented on aiida/cmdline/params/types/path.py:
```diff
@@ -31,3 +33,76 @@ def convert(self, value, param, ctx):
+    def checks_url(self, value, param, ctx):
+        """Do checks for possible URL path"""
+        from six.moves import urllib
+        from django.core.validators import URLValidator
```
Do we really need a django dependency here? I know we have it anyway, but if at some point it needs to be dropped. Isn't there a simply python analogue. Or better, do not validate it all, but literally just go to the next block and try to open it and catch the exception. Maybe the `urllib` library already throws  a specific exception for malformed URLs

CasperWA replied:
I have moved it to the next block and removed URL validation, other than through `urlopen()`.

sphuber commented on aiida/backends/tests/cmdline/commands/test_import.py:
```diff
@@ -85,8 +89,81 @@ def test_comment_mode(self):
+    @unittest.skip("Reenable when issue #2426 has been solved (migrate exported files from 0.3 to 0.4)")
+    def test_import_url_and_local_archives(self):
+        """Test import of both a remote and local archive
+        TODO: Update 'url' to point at correct commit.
```
Is this TODO still relevant and if so what is the 'correct' URL?

CasperWA replied:
I will deepen the explanation in the docstring to include the essence of the following:
It is to be changed when the "migration from v0.3 to v0.4" PR has gone through. Now it is referencing the new export files I created for that PR in Sasha's branch. However, as soon as it has been merged, and Sasha probably deletes his branch, then this URL should point to the, essentially same, commit, but in "aiidateam", not "yakutovicha".

sphuber commented on aiida/cmdline/params/types/path.py:
```diff
@@ -31,3 +33,65 @@ def convert(self, value, param, ctx):
+        super(ImportPath, self).__init__(**kwargs)
+
+        # Set timeout_seconds
+        if timeout_seconds is None:
```
What is the advantage of doing this instead of just putting it in the method signature? The default is immutable so that is fine?

CasperWA replied:
Reiterating a previous comment:

The default value is set in the constructor with a check if timeout_seconds is None set it to the default.
I think it is maybe better to keep a None value in the __init__ parameters, since if the default is set already here and someone should pass a None value the default value will still be used. Unless we specifically want it to raise if a None value is passed?

CasperWA replied:
According to our short talk, I'll simply define `timeout_seconds` in the `__init__` parameters.

CasperWA commented:
Writing some tests for the new class.

CasperWA commented:
Done. New tests for `ImportPath` to test specifically the `timeout_seconds` parameter.
